### PR TITLE
modified the function definitions to allow a custom smtp-from 

### PR DESCRIPTION
**Pull request by ChessSpider**
required for https://github.com/gophish/gophish/pull/986  (see for discussion)

### CODE REVIEW

**Comment by jordan-wright** [comment]
Hey @ChessSpider! Thanks for sending this over.

I appreciate you taking the time to make a first pass at this. Let me review things a bit more, only because changing the function signature of an exported function is a pretty big deal since it breaks backwards incompatibility. It's possible that we could modify either the `Message` or the `Sender` instead to add a parameter that supports the envelope sender.

I'll keep you posted over the next couple of days, as time hopefully frees up for me to work on this. 😄

**Comment by ChessSpider** [comment]
hey Jordan, sure no worries it's your party.  I agree 100% breaking the signature is not a good idea. Probably better to add another function. But I wasn't sure what the best name would be. I was a bit disappointed that Go doesn't support method-overloading. 

I didn't opt for changing the ``Message`` due to it represents the actual Envelope to be sent.

I didn't change the ``Sender`` because there's quite some abstraction-logic there, and it seems to be solely meant for setting up the SMTP-Connection to the server. There's no logic there for sending email as a specific user (as Message is meant for that, but that isn't useful for this use-case)

As the methods used internally already had a From and a To[] which did specifically what I wanted, I chose to do something with that instead.

In any case, have a look and tell me how you'd like to see it. 

tldr; My suggestion is to revert the changed signature back, and add another method which allows for a custom SMTP-From.  

Ideally, the new method also allows for having a different SMTP-To and a different Envelope-To. This means you can set the Envelope-To to e.g. 'undisclosed-recipient' while it still arrives in your inbox.   

Please advise 👍

**Comment by ChessSpider** [comment]
> tldr; My suggestion is to revert the changed signature back, and add another method which allows for a custom SMTP-From.

I have done this now

> Ideally, the new method also allows for having a different SMTP-To and a different Envelope-To. This means you can set the Envelope-To to e.g. 'undisclosed-recipient' while it still arrives in your inbox.

It still cannot do this. But I don't know what the Go-way of approaching this would be.

**Comment by ChessSpider** [comment]
Hi, 
I also snug in automatic generation of message-id per message.. my own anti-spam solution mentioned that as one of the reasons why a message may be spam. 
Because it has to be unique per message I wasn't really sure where else to put it.

File: send.go [hunk 1]
```diff
@@ -5,9 +5,46 @@ import (
...
 	"net/mail"
+	"math/rand"
+	"strings"
 )
 
 // Sender is the interface that wraps the Send method.
+
+
+
+func CreateRandomMessageId(email string) string {
```
Comment (jordan-wright) [inline send.go]: I'd prefer if this could be a separate pull request, since it would require more testing and review. It's easiest to review just one piece of functionality at a time. One way you could accomplish this is to create a new branch off your branch that already has the envelope changes and move the message ID bits there.

Also, if you're looking for a message ID generator, I wrote one for a different email library [here](https://github.com/jordan-wright/email/blob/master/email.go#L580) that we could consider using.

File: smtp_test.go [hunk 2]
```diff
@@ -257,7 +257,7 @@ func testSendMailTimeout(t *testing.T, d *Dialer, want []string) {
 	doTestSendMail(t, d, want, true)
 }
 
-func doTestSMTPReset(t *testing.T, d *Dialer, want []string){
+func doTestSMTPReset(t *testing.T, d *Dialer, want []string) {
```
Comment (jordan-wright) [inline smtp_test.go:260]: If possible, could you add a test to ensure that `SendCustomFrom` is doing what you expect it to do? At the very least, we'll want to make sure that the `MAIL FROM` is what we expect and the `From` header is still the from the message.

**Review by jordan-wright — CHANGES_REQUESTED** [review]
Hey @ChessSpider, this is looking good! It seems that adding the `SendCustomFrom` is about as good as we can do, so that works for me.

I've added a few comments. The biggest thing is that we'll want to move the message ID generation to another separate pull request later down the line.

Thanks again for your work on this. I think we're in the home stretch here! 😄

**Comment by ChessSpider** [comment]
@jordan-wright like this?

**Comment by jordan-wright** [comment]
Looks good to me! Thanks for adding in that test.

**Review by jordan-wright — APPROVED** [review]
